### PR TITLE
Added a new information gain metric for decision tree training.

### PR DESCRIPTION
**Pull request by mhrmm**
@jkinkead Added an alternative "information gain" metric for use during decision tree training.

### CODE REVIEW

**Comment by jkinkead** [comment]
Two tests failed?

File: tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/decisiontree/DecisionTreeTrainer.scala [hunk 1]
```diff
@@ -174,6 +172,7 @@ private class Node(data: Option[FeatureVectorSource], val featureVectorSubset: S
 //param validationPercentage percentage of the training vectors to hold out for validation
```
Comment (jkinkead) [inline tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/decisiontree/DecisionTreeTrainer.scala]: ?

File: tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/decisiontree/DecisionTreeTrainer.scala [hunk 2]
```diff
@@ -229,18 +228,27 @@ class DecisionTreeTrainer(
...
-        // can be expensive
-        val infoGainByFeature: Seq[(Int, Double)] =
-          computeInformationGainUnoptimized(
-            data, node.featureVectorSubset, featuresToExamine
-          ) filter {
-            case (_, gain) => gain > 0 // we get rid of features with zero information gain
-          }
+        // Computes information gain of each selected feature.
+        // Can be expensive.
+        val infoGainByFeature: Seq[(Int, Double)] = informationGainMetric match {
```
Comment (jkinkead) [inline tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/decisiontree/DecisionTreeTrainer.scala]: Case objects are really nice for this:
```
sealed trait GainMetric
case object Entropy extends GainMetric
case object Multinomial extends GainMetric
```
"better than enums"

File: tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/decisiontree/DecisionTreeTrainer.scala [hunk 3]
```diff
@@ -229,18 +228,27 @@ class DecisionTreeTrainer(
...
+            computeEntropyBasedInformationGain(
+              data, node.featureVectorSubset, featuresToExamine
+            ) filter {
+              case (_, gain) => gain > 0 // we get rid of features with low information gain
+            }
+          case "multinomial" =>
+            computeMultinomialBasedInformationGain(
+              data, node.featureVectorSubset, featuresToExamine
+            ) filter {
+              case (_, gain) => gain > 1.0 // we get rid of features with low information gain
```
Comment (jkinkead) [inline tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/decisiontree/DecisionTreeTrainer.scala]: Attach filter function to the gain metric? Or maybe attach threshold to it, if that makes more sense?

File: tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/decisiontree/DecisionTreeTrainer.scala [hunk 4]
```diff
@@ -306,17 +309,37 @@ class DecisionTreeTrainer(
...
-    featureVectorSubset: Seq[Int], featureSubset: Seq[Int]
+    featureVectorSubset: Seq[Int],
+    featureSubset: Seq[Int]
   ): Seq[(Int, Double)] = {
 
-    require(featureVectorSubset.nonEmpty)
-    require(featureSubset.nonEmpty)
+    def computeOutcomeEntropy(featureVectorSubstream: Seq[FeatureVector]) = {
+      val outcomeHistogram: Map[Int, Int] =
+        featureVectorSubstream groupBy { _.outcome.get } mapValues { _.size }
```
Comment (jkinkead) [inline tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/decisiontree/DecisionTreeTrainer.scala]: Why scala doesn't have a `countBy`, I'll never know.

File: tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/decisiontree/DecisionTreeTrainer.scala [hunk 5]
```diff
@@ -306,17 +309,37 @@ class DecisionTreeTrainer(
...
+      computeEntropy(outcomeHistogram)
+    }
 
-    val featureVectorSubstream: Seq[FeatureVector] = featureVectorSubset map {
-      data.getNthVector(_)
+    def computeEntropy(histogram: Map[Int, Int]) = {
```
Comment (jkinkead) [inline tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/decisiontree/DecisionTreeTrainer.scala]: Inline above? Seems weird to have it be a function in the same scope with the same nested level . . .

**Comment by jkinkead** [comment]
A few comments, apart from the failed test.

**Comment by mhrmm** [comment]
Shame on me for not running test before making a pull request.

**Comment by mhrmm** [comment]
Revised!

File: tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/decisiontree/DecisionTreeTrainer.scala [hunk 6]
```diff
@@ -170,10 +168,23 @@ private class Node(data: Option[FeatureVectorSource], val featureVectorSubset: S
   }
 }
 
-/** Functions for training decision trees. */
-//param validationPercentage percentage of the training vectors to hold out for validation
+sealed trait InformationGainMetric {
+  val minimumGain: Double
+}
+case class EntropyGainMetric(minimumGain: Double) extends InformationGainMetric
```
Comment (jkinkead) [inline tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/decisiontree/DecisionTreeTrainer.scala]: If the gain is a constant (like, always `0.0` for entropy), you can also make it a val on the class:
```
case class EntropyGainMetric extends InformationalGainMetric {
  override val minimumGain = 0.0
}
```

Comment (jkinkead) [inline tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/decisiontree/DecisionTreeTrainer.scala]: In that case, you could make these `object` instead of `class`, which saves having to type parens in the `match` expression. ;)

File: tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/decisiontree/DecisionTreeTrainer.scala [hunk 7]
```diff
@@ -229,18 +240,26 @@ class DecisionTreeTrainer(
...
-        // can be expensive
-        val infoGainByFeature: Seq[(Int, Double)] =
-          computeInformationGainUnoptimized(
-            data, node.featureVectorSubset, featuresToExamine
-          ) filter {
-            case (_, gain) => gain > 0 // we get rid of features with zero information gain
-          }
+        // Computes information gain of each selected feature.
+        // Can be expensive.
+        val infoGainByFeature: Seq[(Int, Double)] = (informationGainMetric match {
```
Comment (jkinkead) [inline tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/decisiontree/DecisionTreeTrainer.scala]: Even more fun would be to do:
```
val infoGainByFeature: Seq[(Int, Double)] = (informationGainMetric match {
  case EntropyGainMetric(_) => computeEntropyBaseInformationGain _
  case MultinomialGainMetric(_) => computeMultinomialBasedInformationGain _
})(data, node.featureVectorSubset, featureToExamine) filter {
  case (_, gain) =>
    gain > informationGainMetric.minimumGain // get rid of features with low info gain
}
```

**Comment by jkinkead** [comment]
A couple more distracting comments. Feel free to merge when you're happy!

File: tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/decisiontree/DecisionTreeTrainer.scala (hunk 7 above)
Comment (mhrmm) [inline tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/decisiontree/DecisionTreeTrainer.scala]: That is where I draw the line! :)

File: tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/decisiontree/DecisionTreeTrainer.scala (hunk 6 above)
Comment (mhrmm) [inline tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/decisiontree/DecisionTreeTrainer.scala]: Yeah, but actually this is nice to have as a settable param.

**Comment by mhrmm** [comment]
Thanks!
